src/Npgsql/Replication/PgOutput/ReadonlyArrayBuffer.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using Npgsql.Replication.PgOutput.Messages;
+
+namespace Npgsql.Replication.PgOutput
+{
+    class ReadonlyArrayBuffer<T> : IReadOnlyList<T>
+    {
+        public static readonly ReadonlyArrayBuffer<T> Empty = new(Array.Empty<T>());
+        T[] _items;
+        int _size;
+
+        public ReadonlyArrayBuffer()
+            => _items = Array.Empty<T>();
+
+        ReadonlyArrayBuffer(T[] items)
+        {
+            _items = items;
+            _size = items.Length;
+        }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            for (var i = 0; i < _size; i++)
+            {
+                yield return _items[i];
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public int Count
+        {
+            get => _size;
+            internal set
+            {
+                if (_items.Length < value)
+                    _items = new T[value];
+
+                _size = value;
+            }
+        }
+
+        public T this[int index]
+        {
+            get => index >= _size ? throw new IndexOutOfRangeException() : _items[index];
```

Comment:
Seems like the wrong check... Should be:

```suggestion
            get => index < _size ? _items[index] : throw new IndexOutOfRangeException();
```

Reply:
Yeah, this only inverts it without changing the logic but I agree that it is more intuitive this way.